File: ABC/219/D.py
N = int(input())
X, Y = tuple(map(int, input().split()))
AB = []
for _ in range(N):
    A, B = tuple(map(int, input().split()))
    A = min(A, X)
    B = min(B, Y)
    AB.append([A, B])

# まずAについて貪欲
ans = 0
while True:
    AB = sorted(AB, key=lambda x: (-x[0], -x[1]))
    best = AB[0]
    best_a = best[0]
    best_b = best[1]
    X = max(0, X-best_a)
    Y = max(0, Y-best_b)
    ans += 1
    if (X, Y) == (0, 0):
        print(ans)
        exit()
    if ans == N:
        print(-1)
        exit()
    AB_next = []
    for e in AB[1:]:
        a, b = e[0], e[1]
        a = min(X, a)
        b = min(Y, b)
        AB_next.append([a, b])
    AB = AB_next
    

File: ABC/219/D.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define int long long
#define FOR(i, a, b) for (int i = (a); i < (b); i++)
#define REP(i, n) for (int i = 0; i < (n); i++)
#define pb push_back
#define ALL(obj) (obj).begin(), (obj).end()
#define vi vector<int>
#define vvi vector<vector<int>>
#define Lower_bound(vec, n) distance((vec).begin(), lower_bound((vec).begin(), (vec).end(), (n)))
#define Upper_bound(vec, n) distance((vec).begin(), upper_bound((vec).begin(), (vec).end(), (n)))
//template <class T = int> in()(T x; cin >> x; return (x);)
//template <class T> print(T& x) (cout << x << endl;)
//using template <class T> vec = vector<T>;
#define dual_sort(obj, n) sort((obj).begin(),(obj).end(),[](const vector<int> &alpha,const vector<int> &beta){return alpha[n] < beta[n];})
const int INF = 1LL << 60;
const int MOD = (int)1e9 + 7;
const double EPS = 1e-9;

signed main() {
    int N, X, Y;
    cin >> N >> X >> Y;
    int A, B, S;
    vvi ABS = {};
    REP(i, N) {
        cin >> A >> B;
        A = min(X, A);
        B = min(Y, B);
        S = - A - B;
        ABS.push_back({S, A, B});
    }
    dual_sort(ABS, 0);
    int ans = 0;
    vector<int> best;
    vvi ABS_next;
    vi n_abs;
    while (true) {
        best = ABS[0];
        X = max(0LL, X - best[1]);
        Y = max(0LL, Y - best[2]);
        ans ++;
        if (X == 0LL && Y == 0LL) break;
        if (ans == N) break;
        ABS_next = {};
        REP(i, N-ans) {
            n_abs = ABS[i+1];
            A = n_abs[1];
            B = n_abs[2];
            A = min(X, A);
            B = min(Y, B);
            S = - A - B;
            ABS_next.push_back({S, A, B});
        }
        ABS = ABS_next;
        dual_sort(ABS, 0);
    }
    if (X == 0 and Y == 0) cout << ans << endl;
    else cout << -1 << endl;
}